firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/TaskCompletionSourceCache.java

```diff
@@ -0,0 +1,101 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.appdistribution.impl;
+
+import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.TaskCompletionSource;
+import com.google.firebase.annotations.concurrent.Lightweight;
+import com.google.firebase.concurrent.FirebaseExecutors;
+import java.util.concurrent.Executor;
+
+/**
+ * A cache for a {@link TaskCompletionSource}, for use in cases where we only ever want one active
+ * task at a time for a particular operation.
+ *
+ * <p>This is equivalent to {@link UpdateTaskCache} but for Tasks.
+ */
+class TaskCompletionSourceCache<T> {
+
+  /** A functional interface for a producer of a new {@link TaskCompletionSource}. */
+  @FunctionalInterface
+  interface TaskCompletionSourceProducer<T> {
+
+    /** Produce a new {@link TaskCompletionSource}. */
+    TaskCompletionSource<T> produce();
+  }
+
+  private TaskCompletionSource<T> cachedTaskCompletionSource;
+  private final Executor sequentialExecutor;
+
+  /**
+   * Constructor for a {@link TaskCompletionSourceCache} that controls access using its own
+   * sequential executor backed by the given base executor.
+   *
+   * @param baseExecutor Executor (typically {@link Lightweight}) to back the sequential executor.
```

Comment:
Are there any cases where this should not (or is not) the "lightweight" executor? 

If so, I think it should be documented which circumstances require a different executor. Does it depend on what  `TaskCompletionSourceProducer.produce()` does?

If not, maybe change "typically" to "always" and annotate the consturctor parameter accordingly.